File: Assets/Game/Scripts/Scene/Terrain/TerrainPiece.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class TerrainPiece : BaseSceneObject
{
    /// <summary>
    /// 方向类型
    /// </summary>
    public enum Direction
    {
        Not,    //不可移动
        Flat,   //平面
        Up,     //上升
        Down,   //下沉
    }

    /// <summary>
    /// 空间类型
    /// </summary>
    public enum Space
    {
        Once,   //单一
        Loop,   //循环
    }

    /// <summary>
    ///     ↗z
    ///    /         
    ///   /-----------/→x
    ///  /    |      /
    /// /-----|-----/
    ///       |
    ///      ↓ y
    /// </summary>
    protected Vector3 m_SpacePosition;
    protected Direction m_Direction;
    protected Space m_Space;

    public TerrainPiece(int id, Vector3 spacePosition, Vector3 position, Direction direction, Space space) : base(id,position)
    {
        m_SpacePosition = spacePosition;
        m_Direction = direction;
        m_Space = space;
    }

    public Vector3 GetSpacePosition()
    {
        return m_SpacePosition;
    }

    public Direction GetDirection()
    {
        return m_Direction;
    }

    public Space GetSpace()
    {
        return m_Space;
    }

}

File: Assets/Game/Lua/SceneObject/TerrainPiece/TerrainPiece.lua
TerrainPiece = class(BaseSceneObject)

-- 空间位置坐标图示
--     ↗z
--    /         
--   /-----------/→x
--  /    |      /
-- /-----|-----/
--       |
--       ↓ y 层级

function TerrainPiece:ctor(id, position, spacePosition, direction, space)
    self[BaseSceneObject]:ctor(id, position)

    self.pSpacePosition = spacePosition
    self.pDirection = direction
    self.pSpace = space
end

function TerrainPiece:GetSpacePosition()
    return self.pSpacePosition
end

function TerrainPiece:GetDirection()
    return self.pDirection
end

function TerrainPiece:GetSpace()
    return self.pSpace
end